### DISCUSSION TITLE
Support Service Workers

### DISCUSSION THREAD

**Discussion by radiantly** [violentmonkey/violentmonkey#1219] [Ideas]
[Service workers](https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API) allow you to intercept and modify resources requested by web page. Unfortunately, to register a service worker, one needs to specify an absolute url to a `service-worker.js` that resides under the same origin. I'm curious to know if adding support is feasible in this case.

The flow I have in my mind is:
 - Allow the user to specify a local script to load from a certain URL (eg. `https://originurl.com/service-worker.js`)
 - If a page requests the resource at the above url, intercept and respond with the user's script.
 - With this flow, the user will be able to specify absolute URLs from userscripts.

Of course, this is quite high level, and would love to hear your thoughts on the same - cheers!

**Comment by tophf** [comment]
Userscripts as a concept was created decades ago so in Violentmonkey we don't like extending it unless the proposed feature is already *widely* used by scripts for another userscript engine (most often Tampermonkey) and/or cannot be reimplemented via a polyfill.

Running inside a service worker is something outside of that concept. It requires specialized hacks because you can't use just any URL, it should belong to the target origin. In Firefox it might be possible to patch the original script text via [filterResponseData](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/webRequest/filterResponseData) API but not in Chrome. Hooking up `GM_` functions would be nontrivial either way.

So this looks like something that should be implemented as a separate specialized extension (in Firefox) or maybe even an external utility that proxies the service worker URLs (in Chrome).
